File: advent_2015/elixir/lib/03_santa_visits.ex
defmodule SantaVisits do
  @moduledoc """
  Work out which houses Santa delivers presents to on a grid where there's a
  house at every cell, following instructions to go up|down|left|right.
  Instructions are given as a list of ^ v < > strings.
  """

  @doc """
  Given some input, answer the question in part 1, where we give all
  instructions to Santa, and count the houses he visited at least once.
  """
  def part1(input) do
    input
    |> to_charlist()
    |> visited()
    |> Enum.uniq
    |> Enum.count
  end

  @doc """
  Given some input, answer the question in part 2, where we alternate the
  instructions between Santa and Robo-Santa, and count the houses visited by
  either at least once.
  """
  def part2(input) do
    instructions = to_charlist(input)
    santas = instructions |> Enum.take_every(2)
    robos = instructions |> Enum.drop(1) |> Enum.take_every(2)

    (visited(santas) ++ visited(robos))
    |> Enum.uniq
    |> Enum.count
  end

  @doc """
  Returns a list of unique coordinates (2-tuples) that Santa visited.

  Takes instructions to go up|down|left|right (in the form of ^ v < > codepoints).
  """
  def visited(instructions) do
    Enum.reduce(instructions, [{0,0}], &follow_instructions/2)
  end
  
  defp follow_instructions(?^, [{x, y}|_] = visited), do: [{x, y + 1} | visited]
  defp follow_instructions(?v, [{x, y}|_] = visited), do: [{x, y - 1} | visited]
  defp follow_instructions(?<, [{x, y}|_] = visited), do: [{x + 1, y} | visited]
  defp follow_instructions(?>, [{x, y}|_] = visited), do: [{x - 1, y} | visited]
  defp follow_instructions(_, visited), do: visited
end

File: advent_2015/ruby/03_santa_visits.rb
require 'set'
##
# This class works out which houses Santa delivers presents to, on a grid where
# there's a house at every cell, and instructions to go up|down|left|right.
class SantaVisits
  UP = '^'
  DOWN = 'v'
  LEFT = '<'
  RIGHT = '>'

  def initialize(instructions)
    @instructions = instructions
    fail 'Bad instructions. All characters must be: v^><' if bad_instructions?
  end

  def visited
    Set.new(all_visits)
  end

  private

  attr_reader :instructions

  def all_visits
    instructions.reduce([[0, 0]]) do |visited, instruction|
      visited << new_position(*visited.last, instruction)
    end
  end

  def new_position(x, y, instruction)
    case instruction
    when UP then [x, y + 1]
    when DOWN then [x, y - 1]
    when LEFT then [x - 1, y]
    when RIGHT then [x + 1, y]
    end
  end

  def bad_instructions?
    !instructions.all? { |i| %w(v ^ > <).include? i }
  end
end

##
# Calculate the answer to part 1 and 2
module Answer
  def self.part1(input)
    SantaVisits.new(input.chars).visited.size
  end

  def self.part2(input)
    sliced = input.chars.each_slice(2).to_a
    santas = sliced.map(&:first)
    robos = sliced.map(&:last)

    (SantaVisits.new(santas).visited + SantaVisits.new(robos).visited).size
  end
end

fail unless Answer.part1('>') == 2
fail unless Answer.part1('^>v<') == 4
fail unless Answer.part1('^v^v^v^v^v') == 2

input = File.read('../input.txt').chomp
puts "Santa delivers at least one present to #{Answer.part1(input)} houses."

fail unless Answer.part2('^v') == 3
fail unless Answer.part2('^>v<') == 3
fail unless Answer.part2('^v^v^v^v^v') == 11

puts "Next year, Santa and Robo-Santa deliver to #{Answer.part2(input)} houses."
